gateway-ha/src/main/java/io/trino/gateway/ha/router/QueryCountBasedRouter.java

```diff
@@ -0,0 +1,151 @@
+    //
+    // If a user has queued queries, then find a cluster with the least number of QUEUED queries
+    // for that user.
+    //
+    // If a user's QUEUED query count is the same on every cluster then go with a cluster with
+    // cluster wide stats.
+    //
+    // Find a cluster with the least number of QUEUED queries, if there are the same number of
+    // queries queued, then compare the number of running queries.
+    //
+    // After a query is routed, we need to update the stats for that cluster until we received the
+    // updated stats for all the clusters.
+    // if a user has queries queued then we assume that the routed query will be also queued or
+    // else we assume it would be scheduled immediately and we increment the stats for the running
+    // queries
+
+    private Optional<ClusterStats> getClusterToRoute(String user, String routingGroup)
+    {
+        List<ClusterStats> filteredList = null;
+        Lock readLock = lock.readLock();
+        try {
+            readLock.lock();
+            log.debug("sorting cluster stats for {} {}", user, routingGroup);
+            filteredList = clusterStats.stream()
+                .filter(stats -> stats.healthy())
+                .filter(stats -> routingGroup.equals(stats.routingGroup()))
+                .collect(Collectors.toList());
+        }
+        finally {
+            readLock.unlock();
+        }
+
+        if (filteredList.isEmpty()) {
+            return Optional.empty();
+        }
+
+        Collections.sort(filteredList, new Comparator<ClusterStats>()
+        {
+            public int compare(ClusterStats lhs, ClusterStats rhs)
+            {
+                // First check if the user has any queries queued
+                int compareUserQueue = Integer.compare(lhs.userQueuedCount().getOrDefault(user, 0),
+                        rhs.userQueuedCount().getOrDefault(user, 0));
+
+                if (compareUserQueue != 0) {
+                    return compareUserQueue;
+                }
+
+                int compareClusterQueue = Integer.compare(lhs.queuedQueryCount(),
+                                                            rhs.queuedQueryCount());
+                if (compareClusterQueue != 0) {
+                    return compareClusterQueue;
+                }
+
+                // If the user has equal number of queries queued then see which cluster
+                // has less number of queries running and route it accordingly
+                return Integer.compare(lhs.runningQueryCount(),
+                    rhs.runningQueryCount());
+            }
+        });
+        // The first cluster is right one
+        return Optional.of(filteredList.get(0));
+    }
+
+    private Optional<String> getBackendForRoutingGroup(String routingGroup, String user)
+    {
+        Optional<ClusterStats> cluster = getClusterToRoute(user, routingGroup);
+        if (cluster.isPresent()) {
+            cluster.orElseThrow().updateLocalStats(user);
```

Comment:
Why is `orElseThrow` used? `isPresent()` was just checked. If this is needed then an exception supplier should be passed.

Reply:
That's because the static check doesn't allow it, but I guess there should be a way around it.